gips/0083.md

```diff
@@ -0,0 +1,250 @@
+
+
+## Economic Security and Disputes
+
+Economic security is achieved through slashing and disputes, similar to subgraphs as of Jan 24th 2025. Indexers providing incorrect data CAN have their stake slashed. Therefore, to be an Indexer recognized by the Payment Gateway, one must staked the Network token, just like for subgraphs.
+
+The means are similar to subgraphs, by providing signed attestations, attached to allocations, and having an Arbiter validate those attestations, and slash the Indexer's stake if necessary.
+
+### Attestations
+
+Having no queries like subgraphs have, Proofs of Indexing for Substreams are always null.
+
+> [!NOTE]
+> This is why enabling indexing rewards for Substreams is not part of this GIP. Future GIPs, potentially with Horizon, will propose a mechanism to reward Indexers for providing Substreams services.
+
+Substreams endpoints provide signed attestations with each block's response, as per [the `attestation` field](https://github.com/streamingfast/substreams/blob/develop/proto/sf/substreams/rpc/v2/service.proto#L96) in the `BlockScopedData` response, matching the operator key specified in [the `attestation_public_key` field](https://github.com/streamingfast/substreams/blob/develop/proto/sf/substreams/rpc/v2/service.proto#L106) of the `SessionInit` response.
+
+The attestation payload MUST be signed by a special _attestation key_, derived from the operator's key, so that disputes can be opened on a separate allocation destined for disputes.
+
+A valid _attestation key_, attached to an allocation MUST be verified by the Payment Gateway, before giving the assurance to Consumers that the Indexer can be slashed for misbehaving.
+
+An Arbiter MUST be able to validate payloads, through  multiple providers and investigation, and after judgement, be able to slash the Indexer's stake, using the information provided in those payloads via an on-chain allocations.
+
+
+#### Methods of analysis
+
+THIS SECTION IS NON-NORMATIVE
+
+This section specifies potential methods an Arbiter can use to analyze disputes. Other more optimal tools could be written, and are out of scope of this specification.
+
+**Detecting Faults:** Discrepancies in data returned to a Fisherman or a Consumer, by different providers for the same stream and block, is material to open an investigation.
+
+The Arbiter can use the `substreams gui` tool or the `substreams run --show-attestations` tool to inspect attestations and contents, and compare stream data at specific block ranges, verifying data integrity and determinism.
+
+`substreams tools` will contain a way to validate those attestations, and help with investigation.  This tool will take the content being disputed (BlockScopedData) from the replay.log, which contain the attestations over the content within, and validate their output, in order to compare against providers, and end up slashing an Indexer.
```

Comment:
@abourget , if I understand correctly, the Arbiter would need access to the offended Indexer's replay log, where the attestations live, right? If so, how do we ensure the Indexer always has such a file available should a dispute be opened?

Reply:
Different sinks will consume the stream, so those tools (some are hand-written), would need to keep track of those values. They're part of the stream.

The `substreams gui` has a replay log file format, that outputs in a format `substreams tools` would immediately consume.. It's a simple format and can be replicated (or even hooked in as a middleware for Go-based sinks) easily (it's like length-prefixed `google.protobuf.Any` protobuf messages). If you don't have a sink that tracks it, one could run `substreams gui` for the faulty segment, and save the file.